src/fabric.c

```diff
@@ -327,31 +327,59 @@ static void cleanup_provider(struct fi_provider *provider, void *dlhandle)
 #endif
 }
 
+/* Unknown external provider or duplicate version of a loaded provider.
+ * For duplicate providers, mark all but the newest provider as hidden.
+ * This preserves prior behavior.  Hidden providers can be accessed if
+ * explicitly requested.
+ */
 static struct ofi_prov *ofi_create_prov_entry(const char *prov_name)
 {
-	struct ofi_prov *prov = NULL;
+	struct ofi_prov *prov, *cur, *prev;
+
 	prov = calloc(sizeof *prov, 1);
 	if (!prov) {
 		FI_WARN(&core_prov, FI_LOG_CORE,
 			"Not enough memory to allocate provider registry\n");
 		return NULL;
 	}
 
+	prov->hidden = false;
 	prov->prov_name = strdup(prov_name);
 	if (!prov->prov_name) {
 		FI_WARN(&core_prov, FI_LOG_CORE,
 			"Failed to init pre-registered provider name\n");
 		free(prov);
 		return NULL;
 	}
+
+	for (prev = NULL, cur = prov_head; cur; prev = cur, cur = cur->next) {
+		if ((strlen(prov->prov_name) == strlen(cur->prov_name)) &&
+		    !strcasecmp(prov->prov_name, cur->prov_name)) {
+			if (FI_VERSION_GE(prov->provider->version,
```

Comment:
I think this needs to be reversed? 
HERE: prov is the new provider and cur is the previously loaded one right?
BELOW: prov was the previously loaded one.

Reply:
Might need some comments here.

Reply:
it looks like FI_VERSION_GE it "Greater Than", not "Greater than or Equal to". Normally GE means ">=" and GT means ">".

Reply:
GE is >=.  The check could be GT instead, though the version is the same, so it's a toss-up.

The new provider is newer than the current one, so will be inserted ahead of it in the list (prov->next = cur), and the current one will be marked hidden.  In the else case, the new provider is older.  It is inserted after the current (cur->next = prov) and the new provider is marked hidden.

Reply:
> GE is >=. The check could be GT instead, though the version is the same, so it's a toss-up.
> 
> The new provider is newer than the current one, so will be inserted ahead of it in the list (prov->next = cur), and the current one will be marked hidden. In the else case, the new provider is older. It is inserted after the current (cur->next = prov) and the new provider is marked hidden.

This would be changing the current expected behavior. If 2 providers in the path are the same name, we expect the first one found to be the non-hidden one (or previously the not-loaded one).

Reply:
The current behavior is that only the newer one is kept.  The other is removed.  The new behavior is that the newer one is kept.  The other is hidden.  If you're arguing that the check should be > rather than >=, yes, that's a change and the patch needs to be updated.